### PR TITLE
Factor out attributes as an independent class

### PR DESCRIPTION
**Pull request by aqjune**
Before working on dereferenceable attributes, this factors out attributes as an independent class, because dereferenceable carries additional information (the num of bytes) which cannot be simply encoded as a bit.

### CODE REVIEW

File: ir/memory.cpp [hunk 1]
```diff
@@ -356,18 +356,18 @@ static unsigned bits_shortbid() {
...
-static expr attr_to_bitvec(unsigned attributes) {
+static expr attr_to_bitvec(const ParamAttrs &attrs) {
   if (!bits_for_ptrattrs)
     return expr();
 
   uint64_t bits = 0;
   auto idx = 0;
-  auto to_bit = [&](bool b, Input::Attribute a) -> uint64_t {
-    return b ? (((attributes & a) ? 1 : 0) << idx++) : 0;
+  auto to_bit = [&](bool b, ParamAttrs::Attribute a) -> uint64_t {
```
Comment (nunoplopes) [inline ir/memory.cpp:365]: needs reference here

File: ir/instr.cpp [hunk 2]
```diff
@@ -1271,8 +1271,8 @@ unique_ptr<Instr> InsertValue::dup(const string &suffix) const {
 }
 
 
-void FnCall::addArg(Value &arg, unsigned flags) {
-  args.emplace_back(&arg, flags);
+void FnCall::addArg(Value &arg, const ParamAttrs &attrs) {
```
Comment (nunoplopes) [inline ir/instr.cpp:1274]: this should probably be passed as &&

File: llvm_util/llvm2alive.cpp [hunk 3]
```diff
@@ -271,11 +271,11 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
     unique_ptr<Instr> ret_val;
     auto argI = fn->arg_begin(), argE = fn->arg_end();
     for (auto &arg : args) {
-      unsigned attr = FnCall::ArgNone;
+      ParamAttrs attr = ParamAttrs::None;
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:274]: Maybe introduce an empty constructor

File: ir/value.cpp [hunk 4]
```diff
@@ -188,22 +173,24 @@ StateValue Input::toSMT(State &s) const {
...
-    val = getType().mkInput(s, smt_name.c_str(), attributes);
+    val = getType().mkInput(s, smt_name.c_str(), attrs);
   }
 
   if (!config::disable_undef_input) {
-    auto [undef, vars] = getType().mkUndefInput(s, attributes);
-    for (auto &v : vars) {
-      s.addUndefVar(move(v));
+    auto [undef, vars] = getType().mkUndefInput(s, attrs);
+    if (undef.isValid()) {
```
Comment (nunoplopes) [inline ir/value.cpp:188]: why this change? this is unrelated.
in general, we don't optimize for !isValid() case since most formulas should be valid.

Comment (aqjune) [reply ir/value.cpp:188]: Oh yes, this is an unrelated change. I'll revert this change.
